src/main/java/bio/terra/service/snapshotbuilder/SnapshotAccessRequestModel.java

```diff
@@ -96,37 +110,70 @@ static String generateSummaryForCriteriaGroup(
-                              (SnapshotBuilderDomainCriteria) criteria, settings);
+                              (SnapshotBuilderDomainCriteria) criteria, settings, conceptIdsToName);
                     })
             .collect(Collectors.joining("\n")));
   }
 
   @VisibleForTesting
   static String generateSummaryForCohort(
-      SnapshotBuilderCohort cohort, SnapshotBuilderSettings settings) {
+      SnapshotBuilderCohort cohort,
+      SnapshotBuilderSettings settings,
+      Map<Integer, String> conceptIdsToName) {
     return String.format(
         "Name: %s%nGroups:%n%s",
         cohort.getName(),
         cohort.getCriteriaGroups().stream()
-            .map(criteriaGroup -> generateSummaryForCriteriaGroup(criteriaGroup, settings))
+            .map(
+                criteriaGroup ->
+                    generateSummaryForCriteriaGroup(criteriaGroup, settings, conceptIdsToName))
             .collect(Collectors.joining("\n")));
   }
 
   @VisibleForTesting
-  String generateSummaryFromSnapshotSpecification(SnapshotBuilderSettings settings) {
+  String generateSummaryFromSnapshotSpecification(
+      SnapshotBuilderSettings settings, Map<Integer, String> conceptIdsToName) {
     return snapshotSpecification != null
         ? String.format(
             "Participants included:%n%s%nTables included:%s%n",
             snapshotSpecification.getCohorts().stream()
-                .map(cohort -> generateSummaryForCohort(cohort, settings))
+                .map(cohort -> generateSummaryForCohort(cohort, settings, conceptIdsToName))
                 .collect(Collectors.joining("\n")),
             snapshotSpecification.getOutputTables().stream()
                 .map(SnapshotBuilderOutputTable::getName)
                 .collect(Collectors.joining(", ")))
         : "No snapshot specification found";
   }
 
-  public SnapshotAccessRequestResponse toApiResponse(SnapshotBuilderSettings settings) {
+  @VisibleForTesting
+  List<Integer> generateCriteriaGroupConceptIds(SnapshotBuilderCriteriaGroup criteriaGroup) {
+    return criteriaGroup.getCriteria().stream()
+        // Program data options include the string value in the settings,
+        // so we don't need to fetch their names.
+        .filter(criteria -> criteria.getKind() == SnapshotBuilderCriteria.KindEnum.DOMAIN)
+        .map(criteria -> ((SnapshotBuilderDomainCriteria) criteria).getConceptId())
+        .toList();
+  }
+
+  @VisibleForTesting
+  List<Integer> generateCohortConceptIds(SnapshotBuilderCohort cohort) {
+    return cohort.getCriteriaGroups().stream()
+        .map(this::generateCriteriaGroupConceptIds)
+        .flatMap(List::stream)
+        .toList();
+  }
+
+  public List<Integer> generateConceptIds() {
+    return snapshotSpecification != null
+        ? snapshotSpecification.getCohorts().stream()
+            .map(this::generateCohortConceptIds)
+            .flatMap(List::stream)
+            .distinct()
+            .toList()
+        : List.of();
+  }
```

Comment:
I think it's clearer to write this as a single stream. This also avoids collecting intermediate lists which avoid extra allocation
```suggestion
  public List<Integer> generateConceptIds() {
    return snapshotSpecification.getCohorts().stream()
        .flatMap(cohort -> cohort.getCriteriaGroups().stream())
        .flatMap(criteriaGroup -> criteriaGroup.getCriteria().stream())
        // Program data options include the string value in the settings,
        // so we don't need to fetch their names.
        .filter(criteria -> criteria.getKind() == SnapshotBuilderCriteria.KindEnum.DOMAIN)
        .map(criteria -> ((SnapshotBuilderDomainCriteria) criteria).getConceptId())
        .distinct()
        .toList();
  }
```